pkg/api/policy/checks.go

```diff
@@ -0,0 +1,348 @@
+}
+
+func (this *NotCheck) Exec(target map[string]string,
+	cred map[string]interface{},
+	enforcer Enforcer,
+	currentRule string) bool {
+	return !check(this.rule, target, cred, enforcer, currentRule)
+}
+
+func NewAndCheck(check1 BaseCheck, check2 BaseCheck) *AndCheck {
+	ac := &AndCheck{}
+	ac.AddCheck(check1)
+	ac.AddCheck(check2)
+	return ac
+}
+
+type AndCheck struct {
+	rules []BaseCheck
+}
+
+func (this *AndCheck) String() string {
+	var r []string
+	for _, rule := range this.rules {
+		r = append(r, rule.String())
+	}
+	return fmt.Sprintf("(%s)", strings.Join(r, " and "))
+}
+
+func (this *AndCheck) Exec(target map[string]string,
+	cred map[string]interface{},
+	enforcer Enforcer,
+	currentRule string) bool {
+	for _, rule := range this.rules {
+		if !check(rule, target, cred, enforcer, currentRule) {
+			return false
+		}
+	}
+	return true
+}
+
+func (this *AndCheck) AddCheck(rule BaseCheck) *AndCheck {
+	this.rules = append(this.rules, rule)
+	return this
+}
+
+func NewOrCheck(check1 BaseCheck, check2 BaseCheck) *OrCheck {
+	oc := &OrCheck{}
+	oc.AddCheck(check1)
+	oc.AddCheck(check2)
+	return oc
+}
+
+type OrCheck struct {
+	rules []BaseCheck
+}
+
+func (this *OrCheck) String() string {
+	var r []string
+	for _, rule := range this.rules {
+		r = append(r, rule.String())
+	}
+	return fmt.Sprintf("(%s)", strings.Join(r, " or "))
+}
+
+func (this *OrCheck) Exec(target map[string]string,
+	cred map[string]interface{},
+	enforcer Enforcer,
+	currentRule string) bool {
+	for _, rule := range this.rules {
+		if check(rule, target, cred, enforcer, currentRule) {
+			return true
+		}
+	}
+	return false
+}
+
+func (this *OrCheck) AddCheck(rule BaseCheck) *OrCheck {
+	this.rules = append(this.rules, rule)
+	return this
+}
+
+func (this *OrCheck) PopCheck() (*OrCheck, BaseCheck) {
+	x := this.rules[len(this.rules)-1]
+	this.rules = this.rules[:len(this.rules)-1]
+	return this, x
+}
+
+func NewRuleCheck(kind string, match string) BaseCheck {
+	return &RuleCheck{kind, match}
+}
+
+type RuleCheck struct {
+	kind  string
+	match string
+}
+
+func (this *RuleCheck) String() string {
+	return fmt.Sprintf("%s:%s", this.kind, this.match)
+}
+
+func (this *RuleCheck) Exec(target map[string]string,
+	cred map[string]interface{},
+	enforcer Enforcer,
+	currentRule string) bool {
+	if len(enforcer.Rules) == 0 {
+		return false
+	}
+	return check(enforcer.Rules[this.match], target, cred, enforcer, currentRule)
+}
+
+func keyWorkFormatter(target map[string]string, match string) (string, error) {
+	reg := regexp.MustCompile(`%([[:graph:]]+)s`)
+	if ms := reg.FindAllString(match, -1); len(ms) == 1 {
+		s := ms[0][2 : len(ms[0])-2]
+		for key, val := range target {
+			if s == key {
+				return val, nil
+				break
+			}
+		}
+		return "", fmt.Errorf("target key doesn`t match")
+	}
+	return match, nil
+}
+
+func NewRoleCheck(kind string, match string) BaseCheck {
+	return &RoleCheck{kind, match}
+}
+
+type RoleCheck struct {
+	kind  string
```

Comment:
Is it ok to call `kind` not `Kind`? Because it seems to be the same with parameters in the last method.

Reply:
This is private  variable, so it's ok.